app/app.go

```diff
@@ -0,0 +1,169 @@
+	pki     *pki.PKI
+	auth    *auth.Auth
+	storage *storage.AppStorage
+	server  *server.Server
+
+	cancel context.CancelFunc
+	chErr  chan error
+	chMsg  chan string
+}
+
+func Main() {
+	if err := runApp(context.Background(), newConfig()); err != nil {
+		logs.Error().Err(err).Msg("app exit ERROR")
+		os.Exit(1)
+	}
+
+	logs.Info().Msg("app exit OK")
+	os.Exit(0)
+}
+
+func runApp(ctx context.Context, config *Config) error {
+	app, ctx, err := initApp(ctx, config)
+	if err != nil {
+		return err
+	}
+
+	app.server.Serve(ctx)
+	runtime.Gosched()
+
+	app.testLoad(ctx, 3) //nolint:gomnd
+
+	<-ctx.Done()
+
+	return nil
+}
+
+func (app *App) testLoad(ctx context.Context, xTimes int) { //nolint:revive
+	/* "x-gwi/test/load"
+	go func() {
+		time.Sleep(3 * time.Second) //nolint:gomnd // time to start server
+
+		st, err := load.NewLoadHTTP(ctx, app.server, load.LoadNONE, xTimes) // LoadUA LoadNONE
+		if err != nil {
+			logs.Error().Err(err).Send()
+		}
+
+		// logs.Debug().Msg("closing app after tests")
+		// app.close(ctx)
+		logs.Debug().Interface("load-http", st).Send()
+	}() */
+}
+
+func (app *App) closeApp(ctx context.Context) {
+	if app.cancel != nil {
+		app.cancel()
+	}
+
+	if app.server != nil {
+		app.server.Close(ctx)
+	}
+}
+
+func initApp(ctx context.Context, config *Config) (*App, context.Context, error) {
+	var err error
+	// 1
+	if !config.Valid() {
+		err = fmt.Errorf("invalid config") //nolint:goerr113
+		logs.Error().Err(err).Interface("config", config).Send()
+		context.Background()
+
+		return nil, nil, err
+	}
+	// 2
+	app := &App{ //nolint:exhaustruct
+		config: config,
+		inst:   config.Inst,
+	}
+	if app.inst.Mode() == instance.ModeDev.String() ||
+		app.inst.Mode() == instance.ModeTest.String() {
+		logs.Debug().Interface("config", app.config).Send()
+	}
+	// 3
+	ctx, app.cancel = context.WithCancel(ctx)
+	// 4
+	app.interrupt(ctx)
+	runtime.Gosched()
+	// 5
+	if app.pki, err = pki.NewPKI(); err != nil {
+		return nil, nil, fmt.Errorf("pki.NewPKI: %w", err)
+	}
+	// 6
+	if app.auth, err = auth.NewAuth(ctx, config.Auth, app.inst); err != nil {
+		return nil, nil, fmt.Errorf("auth.NewAuth: %w", err)
+	}
+	// 7
+	if app.storage, err = storage.NewAppStorage(ctx, app.config.Storage, app.inst, app.auth); err != nil {
+		return nil, nil, fmt.Errorf("storage.NewStorage: %w", err)
+	}
+	// 8
+	if app.server, err = server.NewServer(ctx, app.config.Server, app.inst, app.pki); err != nil {
+		return nil, nil, fmt.Errorf("server.NewServer: %w", err)
+	}
+	// 9
+	if err = app.registerServices(app.server.ServiceRegistrar()); err != nil {
+		return nil, nil, fmt.Errorf("app.registerServices: %w", err)
+	}
+	// ok
+	return app, ctx, nil
+}
+
+func (app *App) Valid() bool {
+	return app.config.Valid() &&
+		app.auth.Valid() &&
+		app.storage.Valid() &&
+		app.server.Valid()
+}
+
+func (app *App) interrupt(ctx context.Context) {
+	app.chErr = make(chan error, 2)  //nolint:gomnd
+	app.chMsg = make(chan string, 8) //nolint:gomnd
```

Comment:
Whats with these channels? Why? They are not used.
Why are these channels buffer sizes 2 and 8?

Reply:
Sizes corresponds to potential purpose/behaviour
They are both buffered to not block place of sending in concurrent usage
chErr - is to decide if to close entire app, it is similar to fatal but logic can be modified in interrupt()
chMsg - is more like conditional logging or other central processing of received info; it's usage can be much higher so size  should help with potential peaks

Reply:
OK, thanks. Where would you send data into these channels? Can you provide us with minimal example (just filename/func is enough)?

Reply:
A trivial example but not recommended can be close in app/app.go in func initApp(
by replacement of some immediate return on err to sending this err to app.chErr and continuing
```go
	// 9
	if err = app.registerServices(app.server.ServiceRegistrar()); err != nil {
		//nolint:dupword
		// return nil, nil, fmt.Errorf("app.registerServices: %w", err)
		app.chErr <- err
	}

then
			case err := <-app.chErr:
				// logs.Error().Err(err).Msg("app.chErr")
				// app.closeApp(ctx)
				app.errResolver(ctx, err)
and

func (app *App) errResolver(ctx context.Context, err error) {
	logs.Error().Err(err).Msg("app.chErr")
	app.closeApp(ctx)
}
``` 

For now both channels are consumed by logs in func (app *App) interrupt(
We can enhance by dedicated functions doing some logic with channel taken as an argument

Intended usage is in subroutines like server - grpc but first these channels must be enabled in these scopes in runtime by app.inst which is populated during initApp. And here we have a hierarchy of components so it will be better to provide only direct parent's channel to child(s) to escalate step by step (not directly to highest level)

I ques it is not the case you are asking for, right?
func (app *App) errResolver( doesn't use chan in arguments nor in return values but it is intentional here.
In case there is required chan in arguments or chan in return I know what it is about and how to work with such cases.

btw 
You probably saw the code is organised in kind of hierarchy and avoids global variables and global exported functions. It is to keep consistency of behaviour with flexibility to change the structure from singular to multiple
for example 
current
type App struct {
	server  *server.Server
can become
type App struct {
	server2  []*server.Server
or
type App struct {
	server3  map[instance.InstMode]*server.Server
where server for now = net listener = port x2 plus registered grpc-service-api & gw-mux

In such case a global or wide scope func working with channels can potentially limit flexibility of change 

Let me know, please, if you wish me to enhance the code to enable channels on service level?

Reply:
Thanks, this is fine. The intent is only to pass errors from goroutines created during initialization. Not in the app logic, that was the goal of the questions. Thank you.

Reply:
btw unussed code is now removed or deactivated/commented